xarray/core/common.py

```diff
@@ -679,12 +680,20 @@ def resample(self, freq=None, dim=None, how=None, skipna=None,
 
         if isinstance(dim, basestring):
             dim_name = dim
-            dim = self[dim]
+            dim_array = self[dim]
         else:
             raise TypeError("Dimension name should be a string; "
                             "was passed %r" % dim)
-        group = DataArray(dim, [(dim.dims, dim)], name=RESAMPLE_DIM)
-        grouper = pd.Grouper(freq=freq, closed=closed, label=label, base=base)
+        group = DataArray(dim_array, [(dim_array.dims, dim_array)],
+                          name=RESAMPLE_DIM)
+
+        if isinstance(self.indexes[dim], CFTimeIndex):
+            # TODO: handle closed, label and base arguments, and the case where
+            # frequency is specified without an integer count.
+            grouper = self.indexes[dim].shift(n=int(freq[0]), freq=freq[1:])
```

Comment:
Conveniently, I think this could be written as `self.indexes[dim].shift(1, freq)` and it would handle both the case where the frequency was specified with an integer multiple and without (internally in `shift` the frequency string is converted to an offset that has the appropriate multiple, so `n` can always be 1 in this case):
```
In [1]: import xarray as xr

In [2]: times = xr.cftime_range('2000', periods=5, freq='D')

In [3]: times
Out[3]:
CFTimeIndex([2000-01-01 00:00:00, 2000-01-02 00:00:00, 2000-01-03 00:00:00,
             2000-01-04 00:00:00, 2000-01-05 00:00:00],
            dtype='object')

In [4]: times.shift(1, 'D')
Out[4]:
CFTimeIndex([2000-01-02 00:00:00, 2000-01-03 00:00:00, 2000-01-04 00:00:00,
             2000-01-05 00:00:00, 2000-01-06 00:00:00],
            dtype='object')

In [5]: times.shift(1, '2D')
Out[5]:
CFTimeIndex([2000-01-03 00:00:00, 2000-01-04 00:00:00, 2000-01-05 00:00:00,
             2000-01-06 00:00:00, 2000-01-07 00:00:00],
            dtype='object')
```

Reply:
In general though (e.g. for frequency multiples greater than 1), something as simple as `shift` will not suffice.  I think one will need some sort of binning mechanism like `pandas.cut` to bin the dates in regular intervals starting from the start date, with a result from a call to `xr.cftime_range` forming the bin edges. 

@huard that's all I have for now -- I may add some notes over the weekend if I have time.  I have looked into doing this some in the past, but I didn't arrive at something I was totally pleased with.

Reply:
@spencerkclark Thanks! I'm confused with how groupby on grouper actually works in the example above. I'll take a look at ``cut``.

Reply:
Here's how the magic of `grouper` works:
https://github.com/pydata/xarray/blob/0f70a876759197388d32d6d9f0317f0fe63e0336/xarray/core/groupby.py#L230-L244

Basically, it's just used as the variable over which the groupby operation is done.

Reply:
It's the line ``first_items = s.groupby(grouper).first()`` that feels magic to me. How does ``groubpy`` know to split into years, when ``grouper`` is just a shifted monthly index ?

Reply:
Well, it clearly doesn't work with the shifted monthly index, as Spencer pointed out :).